File: packages/provide-uterm/src/provide/uterm/embed/session.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Embed session, interceptors, and memory upstream."""

from __future__ import annotations

import asyncio
import contextlib
import contextvars
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Protocol

from provide.uterm.embed.types import (
    BackpressurePolicy,
    ByteDirection,
    ClientFilter,
    ClientMetadata,
    DefaultTelnetPolicy,
    InterceptAction,
    InterceptResult,
    SessionLifecycle,
    TelnetPolicy,
    UpstreamPipe,
    WireEventKind,
)

if TYPE_CHECKING:
    from collections.abc import Callable, Mapping

__all__ = [
    "ByteInterceptor",
    "ClientHandle",
    "EmbedSession",
    "InterceptContext",
    "MemoryUpstream",
    "PassThroughInterceptor",
]


@dataclass(slots=True)
class InterceptContext:
    session: EmbedSession
    direction: ByteDirection
    data: bytes
    client_id: str | None = None


class ByteInterceptor(Protocol):
    async def on_upstream(self, context: InterceptContext) -> InterceptResult: ...

    async def on_client(self, context: InterceptContext) -> InterceptResult: ...


class PassThroughInterceptor:
    async def on_upstream(self, context: InterceptContext) -> InterceptResult:
        return InterceptResult.pass_()

    async def on_client(self, context: InterceptContext) -> InterceptResult:
        return InterceptResult.pass_()


class MemoryUpstream:
    """Deterministic test upstream."""

    def __init__(self) -> None:
        self._inbound: asyncio.Queue[bytes | None] = asyncio.Queue()
        self._sent: list[bytes] = []
        self._connected = False
        self._closed = False

    @property
    def is_connected(self) -> bool:
        return self._connected and not self._closed

    @property
    def sent(self) -> list[bytes]:
        return list(self._sent)

    async def connect(self) -> None:
        self._connected = True

    async def disconnect(self) -> None:
        if not self._closed:
            self._closed = True
            await self._inbound.put(None)
        self._connected = False

    async def send(self, data: bytes) -> None:
        if not self.is_connected:
            raise RuntimeError("not connected")
        self._sent.append(bytes(data))

    async def receive(self) -> bytes:
        item = await self._inbound.get()
        if item is None:
            return b""
        return item

    async def push_from_remote(self, data: bytes) -> None:
        if self._closed:
            raise RuntimeError("upstream closed")
        await self._inbound.put(bytes(data))

    def complete_remote(self) -> None:
        self._closed = True
        self._connected = False
        self._inbound.put_nowait(None)


class ClientHandle:
    def __init__(self, client_id: str, meta: ClientMetadata, queue: asyncio.Queue[bytes]) -> None:
        self.client_id = client_id
        self.metadata = meta
        self._queue = queue
        self.is_attached = True

    async def receive(self) -> bytes:
        return await self._queue.get()


@dataclass
class _Deferred:
    direction: ByteDirection
    data: bytes
    client_id: str | None


class EmbedSession:
    """Ordered multi-client proxy session (re-entrancy-safe via single lock)."""

    def __init__(
        self,
        session_id: str,
        *,
        interceptor: ByteInterceptor | None = None,
        telnet_policy: TelnetPolicy | None = None,
        services: Mapping[str, Any] | None = None,
    ) -> None:
        self.session_id = session_id
        self._interceptor: ByteInterceptor = interceptor or PassThroughInterceptor()
        self._telnet = telnet_policy or DefaultTelnetPolicy()
        self.services: dict[str, Any] = dict(services or {})
        self.services["telnet_policy"] = self._telnet
        self.lifecycle = SessionLifecycle.CREATED
        self._upstream: UpstreamPipe | None = None
        self._clients: dict[str, tuple[ClientMetadata, asyncio.Queue[bytes]]] = {}
        self._deferred: list[_Deferred] = []
        self._lock = asyncio.Lock()
        self._reader_task: asyncio.Task[None] | None = None
        self._pipeline_depth: contextvars.ContextVar[int] = contextvars.ContextVar("pipeline_depth", default=0)
        self._on_app: list[Callable[[ByteDirection, bytes, str | None], None]] = []
        self._on_client: list[Callable[[bytes, str | None], None]] = []
        self._on_wire: list[Callable[[WireEventKind, bytes, str], None]] = []
        self._on_life: list[Callable[[SessionLifecycle, str], None]] = []

    def on_application_data(self, cb: Callable[[ByteDirection, bytes, str | None], None]) -> None:
        self._on_app.append(cb)

    def on_client_data(self, cb: Callable[[bytes, str | None], None]) -> None:
        self._on_client.append(cb)

    def on_wire(self, cb: Callable[[WireEventKind, bytes, str], None]) -> None:
        self._on_wire.append(cb)

    def on_lifecycle(self, cb: Callable[[SessionLifecycle, str], None]) -> None:
        self._on_life.append(cb)

    def _set_life(self, phase: SessionLifecycle, detail: str = "") -> None:
        self.lifecycle = phase
        for cb in self._on_life:
            cb(phase, detail)

    async def connect_upstream(self, upstream: UpstreamPipe) -> None:
        async with self._lock:
            original_state = self.lifecycle
            self._set_life(SessionLifecycle.CONNECTING)
            try:
                await upstream.connect()
            except Exception:
                self._set_life(original_state)
                raise
            self._upstream = upstream
            self._start_reader()
            self._set_life(SessionLifecycle.CONNECTED)

    async def replace_upstream(self, upstream: UpstreamPipe) -> None:
        async with self._lock:
            original_state = self.lifecycle
            self._set_life(SessionLifecycle.RECONNECTING)
            old_task = self._reader_task
            self._reader_task = None
            old_up = self._upstream
        if old_task is not None:
            old_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await old_task
        if old_up is not None:
            with contextlib.suppress(Exception):
                await old_up.disconnect()
        async with self._lock:
            self._set_life(SessionLifecycle.CONNECTING)
            try:
                await upstream.connect()
            except Exception:
                self._set_life(original_state)
                raise
            self._upstream = upstream
            self._start_reader()
            self._set_life(SessionLifecycle.CONNECTED)

    async def mark_negotiated(self) -> None:
        async with self._lock:
            self._set_life(SessionLifecycle.NEGOTIATED)
            if self._upstream is not None and self._upstream.is_connected:
                self._set_life(SessionLifecycle.CONNECTED)

    async def attach_client(self, meta: ClientMetadata) -> ClientHandle:
        async with self._lock:
            if not meta.client_id:
                raise ValueError("client_id required")
            if meta.client_id in self._clients:
                raise RuntimeError(f"client already attached: {meta.client_id}")
            cap = max(1, meta.queue_capacity)
            q: asyncio.Queue[bytes] = asyncio.Queue(maxsize=cap)
            self._clients[meta.client_id] = (meta, q)
            self._set_life(SessionLifecycle.CLIENT_ATTACHED, meta.client_id)
            return ClientHandle(meta.client_id, meta, q)

    async def send_to_upstream(self, data: bytes) -> None:
        payload = bytes(data)
        # Re-entrant from interceptor while lock is held (pipeline_depth > 0).
        if self._pipeline_depth.get() > 0:
            await self._process_client(payload, None)
            return
        async with self._lock:
            await self._process_client(payload, None)

    async def send_to_clients(self, data: bytes, filter_: ClientFilter | None = None) -> None:
        async with self._lock:
            f = filter_ or ClientFilter()
            payload = bytes(data)
            self._deliver(payload, f)
            for cb in self._on_app:
                cb(ByteDirection.UPSTREAM_TO_APP, payload, None)

    async def flush_deferred(self) -> None:
        async with self._lock:
            while self._deferred:
                item = self._deferred.pop(0)
                if item.direction is ByteDirection.UPSTREAM_TO_APP:
                    await self._process_upstream(item.data, from_defer=True)
                else:
                    await self._process_client(item.data, item.client_id)

    async def raise_wire(self, kind: WireEventKind, data: bytes, detail: str = "") -> None:
        for cb in self._on_wire:
            cb(kind, bytes(data), detail)

    async def aclose(self) -> None:
        async with self._lock:
            old_task = self._reader_task
            self._reader_task = None
            old_up = self._upstream
            self._upstream = None
            self._clients.clear()
        if old_task is not None:
            old_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await old_task
        if old_up is not None:
            with contextlib.suppress(Exception):
                await old_up.disconnect()
        async with self._lock:
            self._set_life(SessionLifecycle.SHUTDOWN)

    def _start_reader(self) -> None:
        assert self._upstream is not None
        upstream = self._upstream

        async def _loop() -> None:
            try:
                while True:
                    chunk = await upstream.receive()
                    if not chunk:
                        async with self._lock:
                            if self.lifecycle not in (
                                SessionLifecycle.SHUTDOWN,
                                SessionLifecycle.RECONNECTING,
                            ):
                                self._set_life(SessionLifecycle.UPSTREAM_LOST)
                        return
                    async with self._lock:
                        await self._process_upstream(chunk)
            except asyncio.CancelledError:
                return

        self._reader_task = asyncio.create_task(_loop())

    async def _process_upstream(self, data: bytes, *, from_defer: bool = False) -> None:
        token = self._pipeline_depth.set(self._pipeline_depth.get() + 1)
        try:
            ctx = InterceptContext(self, ByteDirection.UPSTREAM_TO_APP, data)
            result = await self._interceptor.on_upstream(ctx)
            await self._apply(result, data, ByteDirection.UPSTREAM_TO_APP, None, from_defer)
        finally:
            self._pipeline_depth.reset(token)

    async def _process_client(self, data: bytes, client_id: str | None) -> None:
        token = self._pipeline_depth.set(self._pipeline_depth.get() + 1)
        try:
            for cb in self._on_client:
                cb(data, client_id)
            ctx = InterceptContext(self, ByteDirection.CLIENT_TO_UPSTREAM, data, client_id)
            result = await self._interceptor.on_client(ctx)
            await self._apply(result, data, ByteDirection.CLIENT_TO_UPSTREAM, client_id, False)
        finally:
            self._pipeline_depth.reset(token)

    async def _apply(
        self,
        result: InterceptResult,
        original: bytes,
        direction: ByteDirection,
        client_id: str | None,
        from_defer: bool,
    ) -> None:
        if result.action is InterceptAction.PASS:
            await self._forward(original, direction)
        elif result.action is InterceptAction.REPLACE:
            await self._forward(result.payload or b"", direction)
        elif result.action is InterceptAction.CONSUME:
            return
        elif result.action is InterceptAction.DEFER:
            if not from_defer:
                self._deferred.append(_Deferred(direction, original, client_id))
        else:
            # INJECT (only remaining InterceptAction)
            payload = result.payload or b""
            if direction is ByteDirection.UPSTREAM_TO_APP:
                await self._process_upstream(payload)
            else:
                await self._process_client(payload, client_id)

    async def _forward(self, data: bytes, direction: ByteDirection) -> None:
        if not data:
            return
        if direction is ByteDirection.UPSTREAM_TO_APP:
            self._deliver(data, ClientFilter())
            for cb in self._on_app:
                cb(direction, data, None)
            return
        if self._upstream is None or not self._upstream.is_connected:
            raise RuntimeError("upstream not connected")
        await self._upstream.send(data)

    def _deliver(self, data: bytes, filter_: ClientFilter) -> None:
        drop: list[str] = []
        for cid, (meta, q) in list(self._clients.items()):
            if not filter_.matches(meta):
                continue
            if not self._try_enqueue(meta, q, data) and meta.backpressure is BackpressurePolicy.DISCONNECT:
                drop.append(cid)
        for cid in drop:
            self._clients.pop(cid, None)

    @staticmethod
    def _try_enqueue(meta: ClientMetadata, q: asyncio.Queue[bytes], data: bytes) -> bool:
        payload = bytes(data)
        if q.full():
            if meta.backpressure is BackpressurePolicy.DROP_NEWEST:
                return False
            if meta.backpressure is BackpressurePolicy.DISCONNECT:
                return False
            # DROP_OLDEST
            with contextlib.suppress(asyncio.QueueEmpty):
                q.get_nowait()
        q.put_nowait(payload)
        return True

File: packages/provide-uterm-csharp/src/Provide.Uterm/Embed/Session.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

using System.Threading.Channels;

namespace Provide.Uterm.Embed;

/// <summary>
/// Ordered multi-client proxy session. All intercept / send paths serialize on a single
/// async gate so re-entrant script output cannot reorder relative to the triggering chunk.
/// </summary>
internal sealed class UtermSession : IUtermSession
{
    private readonly SemaphoreSlim _gate = new(1, 1);
    private readonly AsyncLocal<bool> _holdingGate = new();
    private readonly IByteInterceptor _interceptor;
    private readonly ITelnetPolicy _telnetPolicy;
    private readonly Dictionary<string, object?> _services;
    private readonly Dictionary<string, ClientSlot> _clients = new(StringComparer.Ordinal);
    private readonly Queue<DeferredItem> _deferred = new();
    private readonly object _eventGate = new();

    private IUpstreamPipe? _upstream;
    private CancellationTokenSource? _readerCts;
    private Task? _readerTask;
    private int _disposed;

    private async Task WithGateAsync(Func<Task> action, CancellationToken cancellationToken)
    {
        if (_holdingGate.Value)
        {
            await action().ConfigureAwait(false);
            return;
        }

        await _gate.WaitAsync(cancellationToken).ConfigureAwait(false);
        _holdingGate.Value = true;
        try
        {
            await action().ConfigureAwait(false);
        }
        finally
        {
            _holdingGate.Value = false;
            _gate.Release();
        }
    }

    public UtermSession(string sessionId, EmbedSessionOptions options)
    {
        SessionId = sessionId;
        _interceptor = options.Interceptor ?? PassThroughInterceptor.Instance;
        _telnetPolicy = options.TelnetPolicy ?? new DefaultTelnetPolicy();
        _services = options.Services is null
            ? new Dictionary<string, object?>(StringComparer.Ordinal)
            : new Dictionary<string, object?>(options.Services, StringComparer.Ordinal);
        // Policy is available to hosts via Services convention key.
        _services["telnet_policy"] = _telnetPolicy;
        Lifecycle = SessionLifecycle.Created;
    }

    public string SessionId { get; }
    public SessionLifecycle Lifecycle { get; private set; }
    public IReadOnlyDictionary<string, object?> Services => _services;

    public event EventHandler<ByteChunkEventArgs>? ApplicationDataReceived;
    public event EventHandler<ByteChunkEventArgs>? ClientDataReceived;
    public event EventHandler<WireEventArgs>? WireEvents;
    public event EventHandler<SessionLifecycleEventArgs>? LifecycleChanged;

    public Task ConnectUpstreamAsync(IUpstreamPipe upstream, CancellationToken cancellationToken = default) =>
        WithGateAsync(async () =>
        {
            SetLifecycle(SessionLifecycle.Connecting);
            await upstream.ConnectAsync(cancellationToken).ConfigureAwait(false);
            _upstream = upstream;
            StartReader_NoLock();
            SetLifecycle(SessionLifecycle.Connected);
        }, cancellationToken);

    public async Task ReplaceUpstreamAsync(IUpstreamPipe upstream, CancellationToken cancellationToken = default)
    {
        CancellationTokenSource? oldCts = null;
        Task? oldReader = null;
        IUpstreamPipe? oldUp = null;
        await WithGateAsync(() =>
        {
            SetLifecycle(SessionLifecycle.Reconnecting);
            oldCts = _readerCts;
            oldReader = _readerTask;
            _readerCts = null;
            _readerTask = null;
            oldUp = _upstream;
            return Task.CompletedTask;
        }, cancellationToken).ConfigureAwait(false);

        if (oldCts is not null)
        {
            await oldCts.CancelAsync().ConfigureAwait(false);
        }

        if (oldReader is not null)
        {
            try
            {
                await oldReader.ConfigureAwait(false);
            }
            catch
            {
                // cancelled
            }
        }

        oldCts?.Dispose();
        if (oldUp is not null)
        {
            try
            {
                await oldUp.DisconnectAsync(cancellationToken).ConfigureAwait(false);
            }
            catch
            {
                // best-effort
            }
        }

        await WithGateAsync(async () =>
        {
            SetLifecycle(SessionLifecycle.Connecting);
            await upstream.ConnectAsync(cancellationToken).ConfigureAwait(false);
            _upstream = upstream;
            StartReader_NoLock();
            SetLifecycle(SessionLifecycle.Connected);
        }, cancellationToken).ConfigureAwait(false);
    }

    public Task MarkNegotiatedAsync(CancellationToken cancellationToken = default) =>
        WithGateAsync(() =>
        {
            SetLifecycle(SessionLifecycle.Negotiated);
            if (_upstream?.IsConnected == true)
            {
                SetLifecycle(SessionLifecycle.Connected);
            }

            return Task.CompletedTask;
        }, cancellationToken);

    public Task SendToUpstreamAsync(ReadOnlyMemory<byte> data, CancellationToken cancellationToken = default)
    {
        var bytes = data.ToArray();
        return WithGateAsync(
            () => ProcessClientPath_NoLock(bytes, clientId: null, cancellationToken),
            cancellationToken);
    }

    public Task SendToClientsAsync(ReadOnlyMemory<byte> data, ClientFilter? filter = null, CancellationToken cancellationToken = default)
    {
        var bytes = data.ToArray();
        var f = filter ?? ClientFilter.All;
        return WithGateAsync(() =>
        {
            DeliverToClients_NoLock(bytes, f);
            RaiseApp(bytes, ByteDirection.UpstreamToApp, null);
            return Task.CompletedTask;
        }, cancellationToken);
    }

    public async Task<IClientHandle> AttachClientAsync(ClientAttachOptions options, CancellationToken cancellationToken = default)
    {
        IClientHandle? handle = null;
        await WithGateAsync(() =>
        {
            var meta = options.Metadata;
            if (string.IsNullOrEmpty(meta.ClientId))
            {
                throw new ArgumentException("ClientId required", nameof(options));
            }

            if (_clients.ContainsKey(meta.ClientId))
            {
                throw new InvalidOperationException("client already attached: " + meta.ClientId);
            }

            var slot = new ClientSlot(meta);
            _clients[meta.ClientId] = slot;
            // Event only — do not overwrite durable CONNECTED with ClientAttached.
            LifecycleChanged?.Invoke(this, new SessionLifecycleEventArgs
            {
                Phase = SessionLifecycle.ClientAttached,
                Detail = meta.ClientId,
            });
            handle = slot.Handle;
            return Task.CompletedTask;
        }, cancellationToken).ConfigureAwait(false);
        return handle!;
    }

    public Task DetachClientAsync(string clientId, CancellationToken cancellationToken = default) =>
        WithGateAsync(() =>
        {
            if (_clients.Remove(clientId, out var slot))
            {
                slot.Handle.MarkDetached();
                slot.Complete();
            }

            return Task.CompletedTask;
        }, cancellationToken);

    public Task FlushDeferredAsync(CancellationToken cancellationToken = default) =>
        WithGateAsync(async () =>
        {
            while (_deferred.Count > 0)
            {
                var item = _deferred.Dequeue();
                if (item.Direction == ByteDirection.UpstreamToApp)
                {
                    await ProcessUpstreamPath_NoLock(item.Data, cancellationToken, fromDefer: true).ConfigureAwait(false);
                }
                else
                {
                    await ProcessClientPath_NoLock(item.Data, item.ClientId, cancellationToken, fromDefer: true).ConfigureAwait(false);
                }
            }
        }, cancellationToken);

    public async Task RaiseWireEventAsync(WireEventKind kind, ReadOnlyMemory<byte> data, string detail = "", CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();
        WireEventArgs args;
        lock (_eventGate)
        {
            args = new WireEventArgs { Kind = kind, Data = data.ToArray(), Detail = detail };
        }

        WireEvents?.Invoke(this, args);
        await Task.CompletedTask.ConfigureAwait(false);
    }

    public async ValueTask DisposeAsync()
    {
        if (Interlocked.Exchange(ref _disposed, 1) != 0)
        {
            return;
        }

        CancellationTokenSource? cts = null;
        Task? reader = null;
        IUpstreamPipe? up = null;
        await WithGateAsync(() =>
        {
            cts = _readerCts;
            reader = _readerTask;
            _readerCts = null;
            _readerTask = null;
            up = _upstream;
            _upstream = null;
            foreach (var c in _clients.Values)
            {
                c.Handle.MarkDetached();
                c.Complete();
            }

            _clients.Clear();
            return Task.CompletedTask;
        }, CancellationToken.None).ConfigureAwait(false);

        if (cts is not null)
        {
            await cts.CancelAsync().ConfigureAwait(false);
        }

        if (reader is not null)
        {
            try
            {
                await reader.ConfigureAwait(false);
            }
            catch
            {
                // cancelled
            }
        }

        cts?.Dispose();
        if (up is not null)
        {
            try
            {
                await up.DisconnectAsync().ConfigureAwait(false);
            }
            catch
            {
                // ignore
            }
        }

        await WithGateAsync(() =>
        {
            SetLifecycle(SessionLifecycle.Shutdown);
            return Task.CompletedTask;
        }, CancellationToken.None).ConfigureAwait(false);
        _gate.Dispose();
    }

    private void StartReader_NoLock()
    {
        _readerCts = new CancellationTokenSource();
        var ct = _readerCts.Token;
        var upstream = _upstream ?? throw new InvalidOperationException("no upstream");
        _readerTask = Task.Run(() => ReaderLoopAsync(upstream, ct), CancellationToken.None);
    }

    private async Task ReaderLoopAsync(IUpstreamPipe upstream, CancellationToken ct)
    {
        try
        {
            while (!ct.IsCancellationRequested)
            {
                byte[] chunk;
                try
                {
                    chunk = await upstream.ReceiveAsync(ct).ConfigureAwait(false);
                }
                catch (OperationCanceledException)
                {
                    break;
                }
                catch
                {
                    await OnUpstreamLostAsync().ConfigureAwait(false);
                    break;
                }

                if (chunk.Length == 0)
                {
                    await OnUpstreamLostAsync().ConfigureAwait(false);
                    break;
                }

                await WithGateAsync(
                    () => ProcessUpstreamPath_NoLock(chunk, ct),
                    ct).ConfigureAwait(false);
            }
        }
        catch (OperationCanceledException)
        {
            // normal
        }
    }

    private Task OnUpstreamLostAsync() =>
        WithGateAsync(() =>
        {
            SetLifecycle(SessionLifecycle.UpstreamLost);
            return Task.CompletedTask;
        }, CancellationToken.None);

    private async Task ProcessUpstreamPath_NoLock(byte[] data, CancellationToken ct, bool fromDefer = false)
    {
        var ctx = new InterceptContext
        {
            Session = this,
            Direction = ByteDirection.UpstreamToApp,
            Data = data,
        };
        var result = await _interceptor.OnUpstreamAsync(ctx, ct).ConfigureAwait(false);
        await ApplyResult_NoLock(result, data, ByteDirection.UpstreamToApp, clientId: null, ct, fromDefer).ConfigureAwait(false);
    }

    private async Task ProcessClientPath_NoLock(byte[] data, string? clientId, CancellationToken ct, bool fromDefer = false)
    {
        RaiseClient(data, clientId);
        var ctx = new InterceptContext
        {
            Session = this,
            Direction = ByteDirection.ClientToUpstream,
            Data = data,
            ClientId = clientId,
        };
        var result = await _interceptor.OnClientAsync(ctx, ct).ConfigureAwait(false);
        await ApplyResult_NoLock(result, data, ByteDirection.ClientToUpstream, clientId, ct, fromDefer).ConfigureAwait(false);
    }

    private async Task ApplyResult_NoLock(
        InterceptResult result,
        byte[] original,
        ByteDirection direction,
        string? clientId,
        CancellationToken ct,
        bool fromDefer)
    {
        switch (result.Action)
        {
            case InterceptAction.Pass:
                await Forward_NoLock(original, direction, clientId, ct).ConfigureAwait(false);
                break;
            case InterceptAction.Replace:
                await Forward_NoLock(result.Payload ?? Array.Empty<byte>(), direction, clientId, ct).ConfigureAwait(false);
                break;
            case InterceptAction.Consume:
                break;
            case InterceptAction.Defer:
                if (!fromDefer)
                {
                    _deferred.Enqueue(new DeferredItem(direction, original, clientId));
                }

                break;
            case InterceptAction.Inject:
                // Drop original; re-enter pipeline with inject payload (ordered after current frame completes).
                var injected = result.Payload ?? Array.Empty<byte>();
                if (direction == ByteDirection.UpstreamToApp)
                {
                    await ProcessUpstreamPath_NoLock(injected, ct).ConfigureAwait(false);
                }
                else
                {
                    await ProcessClientPath_NoLock(injected, clientId, ct).ConfigureAwait(false);
                }

                break;
            default:
                await Forward_NoLock(original, direction, clientId, ct).ConfigureAwait(false);
                break;
        }
    }

    private async Task Forward_NoLock(byte[] data, ByteDirection direction, string? clientId, CancellationToken ct)
    {
        if (data.Length == 0)
        {
            return;
        }

        if (direction == ByteDirection.UpstreamToApp)
        {
            DeliverToClients_NoLock(data, ClientFilter.All);
            RaiseApp(data, direction, clientId);
            return;
        }

        var up = _upstream ?? throw new InvalidOperationException("no upstream");
        if (!up.IsConnected)
        {
            throw new InvalidOperationException("upstream not connected");
        }

        await up.SendAsync(data, ct).ConfigureAwait(false);
    }

    private void DeliverToClients_NoLock(byte[] data, ClientFilter filter)
    {
        List<string>? toDrop = null;
        foreach (var (id, slot) in _clients)
        {
            if (!filter.Matches(slot.Metadata))
            {
                continue;
            }

            if (!slot.TryEnqueue(data))
            {
                if (slot.Metadata.Backpressure == BackpressurePolicy.Disconnect)
                {
                    toDrop ??= new List<string>();
                    toDrop.Add(id);
                }
            }
        }

        if (toDrop is not null)
        {
            foreach (var id in toDrop)
            {
                if (_clients.Remove(id, out var slot))
                {
                    slot.Complete();
                }
            }
        }
    }

    private void RaiseApp(byte[] data, ByteDirection direction, string? clientId) =>
        ApplicationDataReceived?.Invoke(this, new ByteChunkEventArgs
        {
            Direction = direction,
            Data = data,
            ClientId = clientId,
        });

    private void RaiseClient(byte[] data, string? clientId) =>
        ClientDataReceived?.Invoke(this, new ByteChunkEventArgs
        {
            Direction = ByteDirection.ClientToUpstream,
            Data = data,
            ClientId = clientId,
        });

    private void SetLifecycle(SessionLifecycle phase, string detail = "")
    {
        Lifecycle = phase;
        LifecycleChanged?.Invoke(this, new SessionLifecycleEventArgs { Phase = phase, Detail = detail });
    }

    private readonly record struct DeferredItem(ByteDirection Direction, byte[] Data, string? ClientId);

    private sealed class ClientSlot
    {
        private readonly Channel<byte[]> _channel;
        private int _count;
        private readonly int _capacity;

        public ClientSlot(ClientMetadata metadata)
        {
            Metadata = metadata;
            _capacity = Math.Max(1, metadata.QueueCapacity);
            _channel = Channel.CreateUnbounded<byte[]>(new UnboundedChannelOptions
            {
                SingleReader = true,
                SingleWriter = false,
            });
            Handle = new ClientHandle(this);
        }

        public ClientMetadata Metadata { get; }
        public ClientHandle Handle { get; }

        public bool TryEnqueue(byte[] data)
        {
            var copy = data.ToArray();
            if (_count >= _capacity)
            {
                switch (Metadata.Backpressure)
                {
                    case BackpressurePolicy.DropNewest:
                        return false;
                    case BackpressurePolicy.Disconnect:
                        return false;
                    case BackpressurePolicy.DropOldest:
                    default:
                        // Best-effort drop one oldest by reading if available.
                        if (_channel.Reader.TryRead(out _))
                        {
                            Interlocked.Decrement(ref _count);
                        }

                        break;
                }
            }

            if (_channel.Writer.TryWrite(copy))
            {
                Interlocked.Increment(ref _count);
                return true;
            }

            return false;
        }

        public async Task<byte[]> ReceiveAsync(CancellationToken ct)
        {
            var item = await _channel.Reader.ReadAsync(ct).ConfigureAwait(false);
            Interlocked.Decrement(ref _count);
            return item;
        }

        public void Complete() => _channel.Writer.TryComplete();
    }

    private sealed class ClientHandle : IClientHandle
    {
        private readonly ClientSlot _slot;
        private volatile bool _attached = true;

        public ClientHandle(ClientSlot slot) => _slot = slot;

        public string ClientId => _slot.Metadata.ClientId;
        public ClientMetadata Metadata => _slot.Metadata;
        public bool IsAttached => _attached;

        internal void MarkDetached() => _attached = false;

        public async Task<byte[]> ReceiveAsync(CancellationToken cancellationToken = default)
        {
            if (!_attached)
            {
                return Array.Empty<byte>();
            }

            try
            {
                return await _slot.ReceiveAsync(cancellationToken).ConfigureAwait(false);
            }
            catch (ChannelClosedException)
            {
                return Array.Empty<byte>();
            }
        }

        public ValueTask DisposeAsync()
        {
            MarkDetached();
            _slot.Complete();
            return ValueTask.CompletedTask;
        }
    }
}
